mono/mini/method-to-ir.c

```diff
@@ -3787,14 +3787,32 @@ mono_emit_load_got_addr (MonoCompile *cfg)
 	MONO_ADD_INS (cfg->bb_exit, dummy_use);
 }
 
+static GENERATE_TRY_GET_CLASS_WITH_CACHE (does_not_return_attr, "System.Diagnostics.CodeAnalysis", "DoesNotReturnAttribute")
+
 static gboolean
 method_does_not_return (MonoMethod *method)
 {
-	// FIXME: Under netcore, these are decorated with the [DoesNotReturn] attribute
-	return m_class_get_image (method->klass) == mono_defaults.corlib &&
-		!strcmp (m_class_get_name (method->klass), "ThrowHelper") &&
-		strstr (method->name, "Throw") == method->name &&
-		!method->is_inflated;
+	// limit to `void Throw*()` methods to reduce expensive checks
+	if (strstr (method->name, "Throw") != method->name)
+		return FALSE;
+
+	if (m_class_get_image (method->klass) == mono_defaults.corlib && 
+		!strcmp (m_class_get_name (method->klass), "ThrowHelper") && !method->is_inflated)
+		return TRUE; // all `Throw*` methods in corlib don't return
+
+	// check for `[DoesNotReturn]` attribute
```

Comment:
DoesNotReturn is not a runtime attribute it seems its used by roslyn for something.

Reply:
@vargaz it's used for null-safety and thus - 100% of corlib will be marked with it eventually where needed, same for user code.
```
The DoesNotReturn it is annotated with will signal to the compiler that no nullable analysis needs to happen after that point, since that code would be unreachable.
```

Reply:
it's defined in corlib but also can be defined anywhere

Reply:
We'll just need to be careful not to remove it. C# compiler flow analysis specific attributes will be removed by linker

Reply:
@marek-safar ah good point